snf-ui/app/components/file-editor.js

```diff
@@ -0,0 +1,80 @@
+import {editFile} from 'snf-ui/snf/editor';
+import {showGroupsOnKeyUp} from 'snf-ui/views/application';
+import {unbindKeyboardShortcuts, bindKeyboardShortcuts} from 'snf-ui/snf/common';
+
+export default Ember.Component.extend({
+
+  didInsertElement: function() {
+    this._super();
+   
+    // remove the X button 
+    $('.close-reveal-modal').remove();
+ 
+    // disable shortcuts
+    unbindKeyboardShortcuts();
+ 
+    
+    // INITIALIZE EDITOR
+    var editor, filename, contents;
+    editor = this.$("#file-editor")[0];
+    filename = this.get('model.name');
+    var readOnly = this.get('readOnly');
+    
+    this.set('contentChanged', false);
+
+    if (!readOnly) {
+      $(".reveal-modal").attr('data-options', 'close_on_background_click:false;close_on_esc:false');
+    }
+
+    var onSave = () => { this.send('save'); }
+    var onChanged = (changed) => { this.set('contentChanged', changed); }
+
+    // reading file requires http request, thus editor  
+    this.set('isLoading', true);
+    this.get('model').read().then((contents) => {
+      this.set('isLoading', false);
+      editFile(editor, filename, contents, readOnly, onSave, onChanged);
+    });
+
+  },
+
+  actions: {
+    save() {
+        if (this.get('contentChanged')) {
+          let contents = editor.getValue();
+          var error500 = false;
+          this.sendAction('onSave', this.get('model'), contents, function(isTrue){
+            if (isTrue === false) {
+              error500 = true;
+            }
+          });
+          if (error500) {
```

Comment:
Okay after some investigation I think I found the problem.

The AJAX API call you create is indeed correct but the response is a bit different.
More specifically, the server returns a **201** status code with an empty response body and the header `Content-Type` set to `application/json`.
![image](https://user-images.githubusercontent.com/13784424/26883306-7042aaf0-4ba5-11e7-9f4f-f3cea1ff3183.png)

In that case jQuery will try to parse the response body as a json. If you try to do so (for example in a console) you will get the following error:

```javascript
JSON.parse("")
Uncaught SyntaxError: Unexpected end of JSON input
    at JSON.parse (<anonymous>)
    at <anonymous>:1:6
```

That's why error handler is always called. [More here](https://stackoverflow.com/questions/14988666/jquery-ajax-done-callback-not-responding-to-201/14988814#14988814)

**A quick and dirty workaround**

Inside the error handler verify the status code and execute the callback accordingly.

**A more clean solution**

As you suggested using the `statusCode` field will be the best option